### PR TITLE
[dynamo] turn torch.backends.cudnn.is_acceptable into a constant

### PR DESCRIPTION
**Pull request by davidberard98**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #90323

Tracing `torch.backends.cudnn.is_acceptable(Tensor) -> bool:` fails with:

```
...
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/variables/functions.py", line 196, in call_function
    return super(UserFunctionVariable, self).call_function(tx, args, kwargs)
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/variables/functions.py", line 67, in call_function
    return tx.inline_user_function_return(
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 426, in inline_user_function_return
    result = InliningInstructionTranslator.inline_call(self, fn, args, kwargs)
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 1698, in inline_call
    return cls.inline_call_(parent, func, args, kwargs)
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 1752, in inline_call_
    tracer.run()
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 485, in run
    and self.step()
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 455, in step
    getattr(self, inst.opname)(inst)
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 281, in wrapper
    return inner_fn(self, inst)
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 912, in CALL_FUNCTION
    self.call_function(fn, args, {})
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/symbolic_convert.py", line 389, in call_function
    self.push(fn.call_function(self, args, kwargs))
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/variables/torch.py", line 431, in call_function
    tensor_variable = wrap_fx_proxy(
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/variables/builder.py", line 662, in wrap_fx_proxy
    return wrap_fx_proxy_cls(
  File "/scratch/dberard/dynamo38/pytorch/torch/_dynamo/variables/builder.py", line 820, in wrap_fx_proxy_cls
    raise AssertionError(
AssertionError: torch.* op returned non-Tensor bool call_function <function is_acceptable at 0x7f00deefb790>
```

So instead, evaluate `is_acceptable()` and convert the result to a constant. The result of `is_acceptable(tensor) -> bool` depends on:
* dtype/device of the input tensor (this should already be guarded)
* properties of the build & whether cudnn is available
* some global state that gets initialized during the first call to `torch.backends.cudnn._init()` (this is NOT guarded in this PR)

Note: this fixes tts_angular with FSDP. This was an issue with FSDP because FSDP modules are interpreted as UnspecializedNNModules, and UnspecializedNNModules try to inline calls. In comparison, NNModules (e.g. when the tts_angular model is not wrapped in FSDP) do not inline calls and instead evaluate subsequent calls. In subsequent calls, cudnn.is_acceptable would be skipped by eval_frame.py:catch_errors because it is not in an allowlist.

cc @mlazos @soumith @voznesenskym @yanboliang @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @chunyuan-w @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

### CODE REVIEW

**Review by wconstab — COMMENTED** [review]
i'm guessing this could be landed as is, but i wonder if instead we can come up with a more generic way to handle cases like these?

can we leverage info in allowlist/skipfiles in any way, or is it ambiguous by not being listed there explicitly?

**Comment by davidberard98** [comment]
Summarizing a discussion I had with @yanboliang:

The failure occurs in [wrap_fx_proxy_cls](https://github.com/pytorch/pytorch/blob/0e182c9441e5702ca1cec4fe34e54f33c38b808b/torch/_dynamo/variables/builder.py#L720), which returns tensor variables, or other special cases that it can handle. We need to avoid calling `wrap_fx_proxy_cls()` on functions that it cannot handle; this includes functions that return scalars (at least [when dynamic shapes are disabled](https://github.com/pytorch/pytorch/blob/0e182c9441e5702ca1cec4fe34e54f33c38b808b/torch/_dynamo/variables/builder.py#L811)).

My question was: why can't we just wrap the call to `wrap_fx_proxy()` in a try-catch block; and if it turns out that we can't handle the function, we can error out? Yanbo mentioned that it's better to explicitly handle these cases to make sure they are handled correctly, because scalar-returning functions like this shouldn't be too common in pytorch.

**Comment by ezyang** [comment]
Hmm, the discussion on this PR seems to be missing the point. Let me try to elucidate.

First, why do Tensor-returning torch API functions "work automatically"? Suppose dynamo traces torch.add(x, y). It doesn't really matter what the internals of add do, because we record a call to this function in the traced graph and expect backends to "rerun" add when we call it on different inputs.

What about torch API functions that don't return Tensor? Well, we only support Tensor (and int, when dynamic shapes is enabled) values in our graph, so we can't actually record the call to this function in the traced graph. And in the case of `is_acceptable`, we are going to use the result of this function call to control what branch of a conditional we go down. So what we need to know is whether or not we can specialize on the output of the function. Whether or not we can specialize depends on the specifics of the function, which is why we can't have a general strategy for this.

If the function is guaranteed to give the same result under the guards we have installed, then we don't need to graph break; we can happily specialize on the result and eliminate the function entirely. If we don't know, it is safest to graph break.

Does that help?

**Comment by yanboliang** [comment]
For the functions returning tensors, how to handle them is very clear: put an op in graph; For the functions don't returning tensors, there are two types:
* We can return the value by dynamo's ```TensorVariable```'s metadata, e.g, [```torch.numel```](https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/variables/torch.py#LL235C9-L240C70) and many others.
* We can't return the value by dynamo's ```TensorVariable```'s metadata ATM. I think graph break to avoid ```Assertion Error``` is okay to workaround before we implement this features by adding necessary metadata.

File: test/dynamo/test_misc.py [hunk 1]
```diff
@@ -2967,6 +2967,34 @@ def fn(x):
         res = opt_fn(x)
         self.assertTrue(same(ref, res))
 
+    @unittest.skipIf(not torch.cuda.is_available(), "requires cuda")
+    @patch.object(torch._dynamo.config, "dynamic_shapes", True)
+    def test_torch_cudnn_is_acceptable_dynamic_shapes(self):
```
[the remark this answers is not in the document]
Comment (davidberard98) [reply test/dynamo/test_misc.py:2972]: @yanboliang I wanted to check that a graph break doesn't happen with dynamic shapes - that's not going to happen automatically right?

**Review by jansel — CHANGES_REQUESTED** [review]
IMO we shouldn't graph break on this op.  We should guard if needed and turn the result into a constant.

File: torch/_dynamo/variables/torch.py [hunk 2]
```diff
@@ -307,9 +310,24 @@ def call_function(
...
             assert len(args) == 2
             return args[1]
+        elif self.value is torch.backends.cudnn.is_acceptable:
+            # is_acceptable(tensor) returns true if
+            #   (a) tensor dtype/device are supported by cudnn
+            #   (b) cudnn is available
+            #   (c) some initialization has completed
+            # technically, it depends on some global state from (c) (torch.backends.cudnn.__cudnn_version)
+            return ConstantVariable(
+                torch.backends.cudnn.is_acceptable(*args, **kwargs), **options
```
Comment (jansel) [inline torch/_dynamo/variables/torch.py:329]: `*args, **kwargs` seems like the wrong thing to pass.  These will be VariableTrackers not pytorch/python types.

File: torch/_dynamo/variables/torch.py [hunk 3]
```diff
@@ -307,9 +310,24 @@ def call_function(
             return NullContextVariable(**options)
         elif self.value is torch.autograd._profiler_enabled:
             unimplemented("torch.autograd._profiler_enabled not supported yet")
+        elif self.unsupported_scalar_return() and not config.dynamic_shapes:
+            # wrap_fx_proxy can only support certain function types; functions that return scalars
+            # are not supported when dynamic shapes are not enabled.
```
Comment (jansel) [inline torch/_dynamo/variables/torch.py:315]: Why should this depend on dynamic shapes?  Dynamo specializes on dtype/device even in dynamic shape mode.

**Review by jansel — CHANGES_REQUESTED** [review]

File: torch/_dynamo/variables/torch.py (hunk 3 above)
Comment (davidberard98) [reply torch/_dynamo/variables/torch.py]: sorry, meant to remove this

File: torch/_dynamo/variables/torch.py (hunk 2 above)
Comment (davidberard98) [reply torch/_dynamo/variables/torch.py:320]: `is_acceptable(tensor)` only needs to use `tensor.device.type` and `tensor.dtype` which are both available on the VariableTracker object; is there a better way to get these? I see there's `get_real_value()` but that seems like overkill if we just want `tensor.device.type` and `tensor.dtype`?

Comment (jansel) [reply torch/_dynamo/variables/torch.py:320]: I'd suggest:
1) Actually check the count and types match, People could do crazy stuff like: `torch.backends.cudnn.is_acceptable("nope", [False])`
2) If the arg is a tensor, we could pass in something like: `torch.tensor(0, dtype=..., device=...)` (assuming all that matters are the dtype/device)

File: test/dynamo/test_misc.py [hunk 4]
```diff
@@ -2967,6 +2967,19 @@ def fn(x):
         res = opt_fn(x)
         self.assertTrue(same(ref, res))
 
+    @unittest.skipIf(not torch.backends.cudnn.is_available(), "requires cudnn")
+    def test_torch_cudnn_is_acceptable(self):
+        def fn(x):
+            if torch.backends.cudnn.is_acceptable(tensor=x):
+                return x.relu()
```
Comment (jansel) [inline test/dynamo/test_misc.py:2974]:
```suggestion
                return x+1
```
will make this test lest flakey.  There is a small chance `x == x.relu()` if x is all positive.

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**